### PR TITLE
Update for azure-chef-extension with separate knife gem

### LINKED ISSUE #346: Install fails with CC17 due to dependency on knife libraries
## Issue
When installing CC17, the install scripts try to require `chef/knife/core/bootstrap_context`

Since `knife` was removed from CC17 this results in a stack trace that looks like this:-

```
:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\1210.13.5.1>ruby C:\Packages\Plugins\Chef.Bootstrap.WindowsAzure.ChefClient\1210.13.5.1\bin\chef-enable.rb
<internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require': cannot load such file -- chef/knife/core/bootstrap_context (LoadError)
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from C:/opscode/chef/embedded/lib/ruby/gems/3.0.0/gems/azure-chef-extension-0.0.1/lib/chef/azure/core/windows_bootstrap_context.rb:19:in `<top (required)>'
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
        from C:/opscode/chef/embedded/lib/ruby/gems/3.0.0/gems/azure-chef-extension-0.0.1/lib/chef/azure/commands/enable.rb:11:in `<top (required)>'
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:160:in `require'
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:160:in `rescue in require'
        from <internal:C:/opscode/chef/embedded/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:149:in `require'
```

## Possible fix
Refactor the bootstrap code to rely on more traditional bootstrap/registration methods such as:-
- Lay down a `client.rb`
- Lay down a `validator.pem`
- Run the chef-client to auto-register
- Remove the `validator.pem` after 1st successful chef run.

### PR DESCRIPTION
**Pull request by ayushbhatt29**
Signed-off-by: ayushbhatt29 <abhatt@msystechnologies.com>

Adding knife gem by directly fetching it from rubygems.org and using the knife libraries from it in place of installed knife gem.

fixes #346

### CODE REVIEW

**Review by NAshwini — CHANGES_REQUESTED** [review]
LGTM, Thank you @ayushbhatt29.
Have few things 1) We need to revert this changes https://github.com/chef-partners/azure-chef-extension/pull/347/files#diff-d7181c68e8be6d24de7da0db9746e17c591232f3941c8e6d400f39b879cb778bR21 and check test case passes.
2) And instead of 
```include Chef::Mixin::ShellOut
ver = shell_out("chef-client -v").stdout
``` 
this can we check by chef-client version by built-in methods directly

**Comment by trickyearlobe** [comment]
This fix wont help customers who don't allow internet access from their Azure accounts (ie. RubyGems access)

Why are we insisting on adding knife back into the equation when Chef client is perfectly able to self register itself using a validator.pem ?

**Comment by ayushbhatt29** [comment]
@trickyearlobe we came up with this solution as we had to support chef15 and chef16 also which work with knife and similarly we added knife by retrieving it from Rubygems for chef-client 17 to use it in windows-bootstrap-context file.

We need to use knife separately for chef-client 17 as there has been architectural change in chef-client from 16 to 17 so knife is removed from inside of chef and is being used separately similar to chef-utils and chef-config.
As windows-bootstrap-context of azure-chef-extension is dependent on bootstrap-context file of knife and enable is dependent on knife that is the reason we require knife for azure chef extension.

The other solution we can work for this is adding knife separately as folder with azure chef extension similar to chef-client 17, to use the required knife files from there. 
This would have a drawback that we will need to update knife folder if its been upgraded.

File: ChefExtensionHandler/bin/chef-install.sh [hunk 1]
```diff
@@ -72,8 +72,8 @@ chef_install_from_script(){
       curl -L -o /tmp/$platform-install.sh https://omnitruck.chef.io/install.sh
       echo "Install.sh script downloaded at /tmp/$platform-install.sh"
       if [ -z "$chef_version" ] && [ -z "$chef_channel" ]; then
-        echo "Installing latest chef-16 client"
-        sh /tmp/$platform-install.sh -v "16" # Until chef-17 is verified
+        echo "Installing latest chef client"
```
Comment (dheerajd-msys) [inline ChefExtensionHandler/bin/chef-install.sh:75]: It can be "chef infra client" here

File: ChefExtensionHandler/bin/chef-install.sh [hunk 2]
```diff
@@ -72,8 +72,8 @@ chef_install_from_script(){
...
+        sh /tmp/$platform-install.sh  
       elif [ ! -z "$chef_version" ] && [ -z "$chef_channel" ]; then
         echo "Installing chef client version $chef_version"
```
Comment (dheerajd-msys) [inline ChefExtensionHandler/bin/chef-install.sh:78]: Same as above ^

File: lib/chef/azure/commands/enable.rb [hunk 3]
```diff
@@ -10,7 +10,18 @@
...
+# Using knife from separately extracted knife gem
+include Chef::Mixin::ShellOut
+ver = shell_out("chef-client -v").stdout
+arr = ver.split(': ')
+if arr[1].start_with?("15") || arr[1].start_with?("16")
+  require 'chef/knife/core/bootstrap_context'
+else
+  find = RUBY_PLATFORM =~ /mswin|mingw|windows/ ? Dir["C:/opscode/chef/knife/*"] : Dir["/opt/chef/knife/*"]
+  path1 = find.join('/')
+  path2 = path1 + '/lib/chef/knife'
```
Comment (dheerajd-msys) [inline lib/chef/azure/commands/enable.rb:22]: Can it be something like 
`path1 = find.join('/') + '/lib/chef/knife'`
`require path1`

File: lib/chef/azure/core/windows_bootstrap_context.rb [hunk 4]
```diff
@@ -16,7 +16,20 @@
...
+include Chef::Mixin::ShellOut
+ver = shell_out("chef-client -v").stdout
+arr = ver.split(': ')
+if arr[1].start_with?("15") || arr[1].start_with?("16")
+  require 'chef/knife/core/bootstrap_context'
+else
+  find = RUBY_PLATFORM =~ /mswin|mingw|windows/ ? Dir["C:/opscode/chef/knife/*"] : Dir["/opt/chef/knife/*"]
+  path1 = find.join('/')
+  path2 = path1 + '/lib/chef/knife/core/bootstrap_context'
+  require path2
```
Comment (dheerajd-msys) [inline lib/chef/azure/core/windows_bootstrap_context.rb:29]: Same as above ^^

File: ChefExtensionHandler/bin/chef-install.sh (hunk 1 above)
Comment (tas50) [reply ChefExtensionHandler/bin/chef-install.sh:75]:
```suggestion
        echo "Installing latest Chef Infra Client"
```

File: ChefExtensionHandler/bin/chef-install.sh (hunk 2 above)
Comment (tas50) [inline ChefExtensionHandler/bin/chef-install.sh:78]:
```suggestion
        echo "Installing Chef Infra Client version $chef_version"
```

File: azure-chef-extension.gemspec [hunk 5]
```diff
@@ -18,7 +18,7 @@ Gem::Specification.new do |s|
   s.executables   = `git ls-files -- bin/*`.split("\n").map{ |f| File.basename(f) }
   s.require_paths = ["lib", "spec"]
 
-  s.add_development_dependency "chef", '<= 16.13.16'
+  s.add_development_dependency "chef"
```
Comment (tas50) [inline azure-chef-extension.gemspec:21]: You probably still want to pin here

**Comment by tas50** [comment]
Why are we not just depending on knife in the gemspec? We should not be extracting parts of knife. That's going to end poorly for us.

**Comment by trickyearlobe** [comment]
My earlier question still stands... why are we installing knife at all when the original reasons for removing it were `Customers don't want the security risk of having admin tools on their VM's` and `removing knife makes the client smaller`.

When we install using the Azure Extension, we have a `validator.pem` and all the versions of Chef Client I've ever used (going back to at least 2013) know how to register themselves using a `validator.pem`, without knife.

We have an opportunity here to clean up the code quite a bit. We've already had fragility issues caused by assumptions that the internals of knife won't change and that its safe to ship bits of code from earlier versions and use them with latest Chef Client (see https://github.com/chef-partners/azure-chef-extension/pull/308/commits/a688b6f1c06e8a956a343db3e8e5bae518dffc7a) as an example.

I feel like the Azure Extension breaks pretty much every time we ship a new major version of Chef Client.

**Comment by tas50** [comment]
If we can avoid using knife for a bootstrap entirely that would be preferred, but we 100% cannot be extracting portions of the knife gem for use. That's just a hack and one that will break in terrible, customer impacting, ways.

**Comment by ayushbhatt29** [comment]
> Why are we not just depending on knife in the gemspec? We should not be extracting parts of knife. That's going to end poorly for us.

@tas50 I have tested using knife as a dependency in gemspec and gemfile also, but while setting the extension, knife gem is not getting installed with it and is giving `knife gem not found` error.
This is why i fetched the knife gem directly from rubygems.org as installing knife would require dev-tools installed on the vm where we want to add extension.

File: ChefExtensionHandler/bin/chef-install.sh (hunk 2 above)
Comment (ayushbhatt29) [reply ChefExtensionHandler/bin/chef-install.sh]: Thanks @dheerajd-msys

Comment (ayushbhatt29) [reply ChefExtensionHandler/bin/chef-install.sh]: Thanks @tas50

File: azure-chef-extension.gemspec (hunk 5 above)
Comment (ayushbhatt29) [reply azure-chef-extension.gemspec:21]: This pin was added when we were pinning the azure chef extension to Chef-Infra-Client 16, so to removing the pin for latest Chef Infra Client this pin will need to removed from here.

**Comment by ayushbhatt29** [comment]
I have also tried with using knife as a non-dev dependency in gemspec and also as a run-time dependency but that is also not installing the knife with azure-chef-extension.

Alternatively for using knife we can add the complete knife as a separate directory with azure chef extension code so we can use it directly from there.

**Comment by ayushbhatt29** [comment]
Closing this PR as [PR 358](https://github.com/chef-partners/azure-chef-extension/pull/358) has solution for this issue by removing knife dependency.

**Pull request closed without merging**